### PR TITLE
use bearssl rng throughout

### PR DESCRIPTION
**Pull request by arnetheduck**

### CODE REVIEW

**Comment by arnetheduck** [comment]
> is now avoided.

strictly, it's moved from being spread out all across the codebase to a single location - the newRng function

> technical benefit

the main objective is to use a single method across the codebase, and bear happened to satisfy that criteria and already be included. technically, it could be wrapped in our own type / interface if we wanted to isolate it further (like `crypto.nim` does for bls in eth2) but this seemed onerous.

> terms of security

https://bearssl.org/gitweb/?p=BearSSL;a=blob;f=inc/bearssl_rand.h;h=0a9f544fc4c58a82805d9b759eb6b21d8a001511;hb=dda1f8a0c46e15b4a235163470ff700b2f13dcc5#l76

> never need to reseed

see above - also, since it's a single instance, it's easy to reseed periodically

**Comment by kdeme** [comment]
> it could be wrapped in our own type / interface if we wanted to isolate it further

Yes, was about to suggest that but then didn't as I realized we would still depend on bearssl in nim-eth for using it in `dcli` or in the tests (if we want to set proper example). Nevertheless, it might still be a sound good idea, we already use it in nbc, but we might to not want to push to other projects.  But then again, those projects would also need a good rng, so might end up with bearssl again :).

File: eth/keys.nim [hunk 1]
```diff
@@ -46,12 +46,29 @@ type
...
 
-proc random*(T: type PrivateKey): SkResult[T] =
-  SkSecretKey.random().mapConvert(T)
+template pubkey*(v: KeyPair): PublicKey = PublicKey(SkKeyPair(v).pubkey)
+template seckey*(v: KeyPair): PrivateKey = PrivateKey(SkKeyPair(v).seckey)
+
+proc newRng*(): ref BrHmacDrbgContext =
+  # You should only create one instance of the RNG per application / library
+  # Ref is used so that it can be shared between components
+  # TODO consider moving to bearssl
```
Comment (kdeme) [inline eth/keys.nim:57]: Yeah, I think that would be better. Looking at the nbc code now that calls this, it looks weird and confusing

File: eth/keys.nim [hunk 2]
```diff
@@ -46,12 +46,29 @@ type
...
+  var
+    seeder = brPrngSeederSystem(nil)
```
Comment (kdeme) [inline eth/keys.nim:59]: Looks like this could also fail in theory

File: eth/keys.nim (hunk 1 above)
Comment (arnetheduck) [reply eth/keys.nim:57]: yeah - I've thought of refactoring bearssl a little, would do it then

**Comment by arnetheduck** [comment]
> good rng

yes - and at the core of the matter lies that some of our key generation is tied to the bear rng explicitly so until that is rewritten, this is a freedom we can't have - I'd leave it for another day

**Comment by arnetheduck** [comment]
> RNG in a high-level API

yes - different PR though - one thing to consider is that bear functions internally depend on this being the bear API specifically

**Pull request merged**